ZOOKEEPER-2967: Add check to validate dataDir and dataLogDir paramete…

mfenes opened the pull request:
ZOOKEEPER-2967: Add check to validate dataDir and dataLogDir parameters at startup

This PR adds a check to protect ZK against configuring dataDir and dataLogDir opposingly.

When FileTxnSnapLog is created, it checks if transaction log directory contains snapshot files or vice versa, snapshot directory contains transaction log files. If so, the check throws LogdirContentCheckException or SnapdirContentCheckException, respectively, which translates to DatadirException at ZK startup in QuorumPeerMain and ZooKeeperServerMain.

If the two directories are the same, then no check is done.

For testing, I've added 4 new unit tests which cover the following cases:

1. transaction log and snapshot directories are different and they are used correctly (no Exception)
2. transaction log and snapshot directories are the same (in this case no check is done)
3. transaction log and snapshot directories are different and transaction log directory contains snapshot files (LogdirContentCheckException -> ZK quits)
4. transaction log and snapshot directories are different and snapshot directory contains transaction log files (SnapdirContentCheckException -> ZK quits)

anmolnar commented on src/java/main/org/apache/zookeeper/server/persistence/Util.java:
```diff
@@ -83,7 +86,7 @@ public static URI makeFileLoggerURL(File dataDir, File dataLogDir,String convPol
      */
     public static String makeLogName(long zxid) {
-        return "log." + Long.toHexString(zxid);
+        return LOG_FILE_PREFIX + "." + Long.toHexString(zxid);
```
I wonder if LOG_FILE_PREFIX included the dot, would be better for the validation.

mfenes replied:
I deliberately excluded the dot from LOG_FILE_PREFIX and SNAP_FILE_PREFIX after I checked that in FileTxnLog and FileSnap classes the prefix argument (like "log", "snapshot") in method calls is passed without the dot. I wanted to avoid confusion.

anmolnar replied:
Okay.

anmolnar commented on src/java/main/org/apache/zookeeper/server/persistence/Util.java:
```diff
@@ -294,5 +297,25 @@ public int compare(File o1, File o2) {
+     * @return
+     */
+    public static boolean isLogFile(File file) {
+        return file.getName().startsWith(LOG_FILE_PREFIX);
```
It might be safer to check here with the dot (".") included in the prefix.

mfenes replied:
I agree that it might be safer to check with the dot. I'll add this change.

mfenes replied:
This is done.

anmolnar commented on src/java/main/org/apache/zookeeper/server/persistence/Util.java:
```diff
@@ -294,5 +297,25 @@ public int compare(File o1, File o2) {
+     * @return
+     */
+    public static boolean isSnapshotFile(File file) {
+        return file.getName().startsWith(SNAP_FILE_PREFIX);
```
Same here with the dot.

mfenes replied:
Ok, I'll include the dot in the check.

mfenes replied:
Done.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+        File logVersionDir = new File(logDir, FileTxnSnapLog.version +  FileTxnSnapLog.VERSION);
+        File snapVersionDir = new File(snapDir, FileTxnSnapLog.version +  FileTxnSnapLog.VERSION);
+
+        if (!logVersionDir.exists()) {
```
This is always true.

mfenes replied:
I've removed this.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+        if (!logVersionDir.exists()) {
+            logVersionDir.mkdirs();
+        }
+        if (!snapVersionDir.exists()) {
```
This is always true.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+        }
+
+        Assert.assertTrue(logVersionDir.exists());
+        Assert.assertTrue(snapVersionDir.exists());
```
These assert will never fail. mkdirs() throws exception if the directory cannot be created.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+        Assert.assertTrue(logFile1.exists());
+        Assert.assertTrue(logFile2.exists());
+        Assert.assertTrue(snapFile1.exists());
+        Assert.assertTrue(snapFile2.exists());
```
Again, I believe these checks are redundant. createNewFile() will always successfully create the file or throws exception.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+            fileTxnSnapLog = new FileTxnSnapLog(logDir, snapDir);
+        } catch (FileTxnSnapLog.LogdirContentCheckException e) {
+            Assert.fail("Should not throw LogdirContentCheckException.");
+        } catch (FileTxnSnapLog.SnapdirContentCheckException e) {
```
These 2 catches can be grouped by catching all Exceptions.

mfenes replied:
Done.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+                System.clearProperty(FileTxnSnapLog.ZOOKEEPER_DATADIR_AUTOCREATE);
+            } else {
+                System.setProperty(FileTxnSnapLog.ZOOKEEPER_DATADIR_AUTOCREATE, priorAutocreateDirValue);
+            }
```
Setting 'zookeeper.datadir.autocreate' doesn't make any difference here, so it doesn't need to be maintained.
In addition to that please add clean-up logic to your tests. (e.g. recursively remove temp directories with all their contents)

mfenes replied:
I switched off autocreate to make sure the test is reading the test input data and not something else by mistake.
Added clean-up of temp directories.

anmolnar commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
+    }
+
+    @Test
+    public void testDirCheckWithSameLogAndSnapDirs() throws IOException {
```
I don't repeat my previous comments, all of them applies to the rest of test cases.

mfenes commented on src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java:
```diff
@@ -136,13 +136,51 @@ public FileTxnSnapLog(File dataDir, File snapDir) throws IOException {
     }
 
+    private void checkLogDir() throws LogdirContentCheckException {
+        File[] files = this.dataDir.listFiles();
```
If I used FilenameFilter, then Util.isSnapshotFile() / Util.isLogFile() check would be run for all the files in the directory and listFiles(FilenameFilter filter) would return all the files satisfying the filter condition, however I need only the first occurrence which satisfies the condition, not all of them. The current logic quits from the for loop immediately when it finds a file violating the configuration and throws an exception, while your proposal would iterate over all the files in the directory and would call Util.isSnapshotFile() / Util.isLogFile() for each of the files inside FilenameFilter to prepare the filtered File[]. So using FilenameFilter would be a bit slower, but yes, it might need less lines in code, also at the price of obscuring the purpose of the code (i.e. hasSnapshotFiles / hasLogFiles boolean variables tell what the problem exactly is, while if (snapshotFiles.length > 0) { throw new Exception(...) } would not). However, if we prefer using Java library classes over standard coding patterns even in cases when it does not fit the purpose entirely, then FilenameFilter can be the winner.

anmolnar replied:
I prefer writing (and later reading) less code than more. You could be right about the performance impact, but at least `listFiles()` will only create File objects for the matching files, not for all of them.

mfenes replied:
Changed implementation to use FilenameFilter.

mfenes commented on src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java:
```diff
@@ -136,13 +136,51 @@ public FileTxnSnapLog(File dataDir, File snapDir) throws IOException {
+    }
+
+    private void checkSnapDir() throws SnapdirContentCheckException {
+        File[] files = this.snapDir.listFiles();
```
Please see my comments above.

afine commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -159,4 +159,222 @@ public void onTxnLoaded(TxnHeader hdr, Record rec) {
         }
     }
+
+    @Test
```
there is a lot of code duplication in these tests, i'm wondering if they can be cleaned up.

mfenes replied:
@afine I've improved/shortened the test code

mfenes commented:
@afine I've made the following changes:
Replaced "log" and "snapshot" strings with constants FileTxnLog.LOG_FILE_PREFIX and FileSnap.SNAPSHOT_FILE_PREFIX.
Used Util.makeLogName when creating new transaction log file to make sure creation and verification follows the same file name pattern. 
Added reference to ZOOKEEPER-2967 to explain why this check is needed.
Cleaned up/refactored unit test FileTxnSnapLogTest to make it less verbose and consistent.

mfenes commented on src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java:
```diff
@@ -136,13 +136,51 @@ public FileTxnSnapLog(File dataDir, File snapDir) throws IOException {
             throw new DatadirException("Cannot write to snap directory " + this.snapDir);
         }
 
+        // check content of transaction log and snapshot dirs if they are two different directories
```
Added reference to ZOOKEEPER-2967 in the comments.

eolivelli commented on src/java/test/org/apache/zookeeper/server/persistence/FileTxnSnapLogTest.java:
```diff
@@ -35,92 +38,181 @@
+
+    @Before
+    public void setUp() throws Exception {
+        tmpDir = ClientBase.createEmptyTestDir();
```
Maybe it could be a good improvement to start using junit TemporaryFolder rule so that the clean up will ve automatically handled by junit

mfenes replied:
@eolivelli Thanks for the idea. I'll check and use this next time.
